src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+            !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("RobinDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow(e)
+        end
+    end
+end
+
+hash_key(key) = ((hash(key)%UInt32) & 0xffffffff)|1
+desired_index(hash, sz) = ((hash) & (sz -1)) + 1
+
+function calculate_distance(h::RobinDict{K, V}, index) where {K, V} 
+    @assert isslotfilled(h, index)
+    sz = length(h.keys)
+    @inbounds index_init = desired_index(h.hashes[index], sz)
+    return (index - index_init + sz) & (sz - 1)
+end
+
+# insert algorithm
+function rh_insert!(h::RobinDict{K, V}, key::K, val::V) where {K, V}
+    # table full
+    @assert h.count != length(h.keys)
+    
+    ckey, cval, chash = key, val, hash_key(key)
+    sz = length(h.keys)
+    index_init = desired_index(chash, sz)
+
+    index_curr = index_init
+    probe_distance = 0
+    probe_current = 0
+    @inbounds while true
+        if (isslotempty(h, index_curr)) || (isslotfilled(h, index_curr) && isequal(h.keys[index_curr], ckey))
+            break
+        end
+        probe_distance = calculate_distance(h, index_curr)
+
+        if probe_current > probe_distance
+            h.vals[index_curr], cval = cval, h.vals[index_curr]
+            h.keys[index_curr], ckey = ckey, h.keys[index_curr]
+            h.hashes[index_curr], chash = chash, h.hashes[index_curr]
+            probe_current = probe_distance
+        end
+        probe_current += 1
+        index_curr = (index_curr & (sz - 1)) + 1
+    end
+    
+    @inbounds if isslotfilled(h, index_curr) && isequal(h.keys[index_curr], ckey)
+        h.vals[index_curr] = cval
+        return index_curr
+    end
+
+    @inbounds if isslotempty(h, index_curr)
+        h.count += 1
+    end
+
+    @inbounds h.vals[index_curr] = cval
+    @inbounds h.keys[index_curr] = ckey
+    @inbounds h.hashes[index_curr] = chash
+    
+    @assert probe_current >= 0
+    
+    h.maxprobe = max(h.maxprobe, probe_current)
+    if h.idxfloor == 0
+        h.idxfloor = index_curr
+    else
+        h.idxfloor = min(h.idxfloor, index_curr)
+    end
+    return index_curr
+end
+
+#rehash! algorithm
+function rehash!(h::RobinDict{K,V}, newsz = length(h.keys)) where {K, V}
+    oldk = h.keys
+    oldv = h.vals
+    oldh = h.hashes
+    sz = length(oldk)
+    newsz = _tablesz(newsz)
+    h.totalcost += 1
+    if h.count == 0
+        resize!(h.keys, sz)
+        resize!(h.vals, sz)
+        resize!(h.hashes, newsz)
+        fill!(h.hashes, 0)
+        h.count = 0
+        h.maxprobe = 0
+        h.totalcost = 0
+        h.idxfloor = 0
+        return h
+    end
+
+    h.keys = Vector{K}(undef, newsz)
+    h.vals = Vector{V}(undef, newsz)
+    h.hashes = zeros(UInt32,newsz)
+    totalcost0 = h.totalcost
+    h.count = 0
+    h.maxprobe = 0
+    h.idxfloor = 0
+
+    for i = 1:sz
+        @inbounds if oldh[i] != 0
+            k = oldk[i]
+            v = oldv[i]
+            rh_insert!(h, k, v)
```

Comment:
`rehash!` should not compute a single hash; instead we should reuse the stored hashes.

Reply:
So, that would require an alternate `insert_for_rehash`, right?

Reply:
Exactly. `Base` also has an alternate `insert_for_rehash` that never calls `isequal` on keys; here, we could additionally save the `hash` calls.